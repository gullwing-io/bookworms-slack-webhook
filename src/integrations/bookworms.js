import { loadBookmarks, generateBookmarks } from "bookworms";

let bookmarks = "";

const init = async (path) => {
  const { body } = await loadBookmarks.fetchBookmarkConfig(path);
  bookmarks = body;
};

// generating the markdown only for the top level folder requested in the slash command
const generateBookmarkMarkdown = (folder) => {
  const requestedFolder = findBookMarkFolder(folder, bookmarks.folders);
  if (requestedFolder) {
    const selectedBookmarks = {
      folders: [requestedFolder],
    };
    const [browser, readme] =
      generateBookmarks.generateImportBookmarkMarkup(selectedBookmarks);
    // taking the readme as this is markdown
    return readme.body;
  } else {
    return `Sorry ${folder} could not be found, use \`/bookmarks all\` for a list of available bookmarks    `;
  }
};

// used for showing a dynamic list of available top level folders
const listOfBookmarksMarkDown = () => {
  const commands = bookmarks.folders.map((folder) => {
    return `\`/bookmarks ${folder.label.toLowerCase()}\``;
  });
  // adding in new line for formatting in Slack response
  const commandsMarkdown = `
${commands.join("\r\n")}
    `;
  return `Bookmarks are seperated into different domains, to get the specific bookmarks you can enter the following commands
    ${commandsMarkdown}
    `;
};

// returning the markdown of the requested top level folder
const findBookMarkFolder = (name, folders) => {
  return folders.find((folder) => {
    if (folder.label.toLowerCase() === name) {
      return folder;
    }
  });
};

const getBookmarks = (folder) => generateBookmarkMarkdown(folder);

export { init, getBookmarks, listOfBookmarksMarkDown, findBookMarkFolder };
