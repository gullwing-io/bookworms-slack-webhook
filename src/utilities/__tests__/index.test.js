import { jest } from "@jest/globals";
import { init } from "../../integrations/bookworms.js";
import { requestHandler, defaultRoute } from "../index.js";

// I rather not test against real data but mocks with native imports are causing me issues
await init(
  "https://raw.githubusercontent.com/thearegee/bookworms/main/demo/config/bookmarks.yaml"
);

describe("Utilties", () => {
  describe("requestHandler", () => {
    test("should return valid folder found", () => {
      const body = {
        text: "folder 1",
      };
      const res = {
        send: jest.fn(),
      };
      requestHandler(body, res);
      expect(res.send).toMatchSnapshot();
    });
    test("should return folder not found", () => {
      const body = {
        text: "junk",
      };
      const res = {
        send: jest.fn(),
      };
      requestHandler(body, res);
      expect(res.send).toMatchSnapshot();
    });
  });
  test("should return all bookmarks - no text", () => {
    const body = {};
    const res = {
      send: jest.fn(),
    };
    requestHandler(body, res);
    expect(res.send).toMatchSnapshot();
  });
  test("should return all bookmarks - blank text", () => {
    const body = {
      text: "",
    };
    const res = {
      send: jest.fn(),
    };
    requestHandler(body, res);
    expect(res.send).toMatchSnapshot();
  });
  test("should return all bookmarks - all text", () => {
    const body = {
      text: "",
    };
    const res = {
      send: jest.fn(),
    };
    requestHandler(body, res);
    expect(res.send).toMatchSnapshot();
  });
  describe("defaultRoute", () => {
    test("should return default route", () => {
      expect(defaultRoute).toEqual("/webhooks/slack/bookworms");
    });
  });
});
