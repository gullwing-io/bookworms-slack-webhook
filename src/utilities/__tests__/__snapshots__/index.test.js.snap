// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utilties requestHandler should return folder not found 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "blocks": Array [
          Object {
            "text": Object {
              "text": "Bookmarks",
              "type": "plain_text",
            },
            "type": "header",
          },
          Object {
            "type": "divider",
          },
          Object {
            "text": Object {
              "text": "Sorry junk could not be found, use \`/bookmarks all\` for a list of available bookmarks
",
              "type": "mrkdwn",
            },
            "type": "section",
          },
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Utilties requestHandler should return valid folder found 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "blocks": Array [
          Object {
            "text": Object {
              "text": "Bookmarks",
              "type": "plain_text",
            },
            "type": "header",
          },
          Object {
            "type": "divider",
          },
          Object {
            "text": Object {
              "text": "*folder 1*

This is to describe the folder structure

*sub folder 1*

This is to describe the sub folder structure

•   <https://www.mywebsite.com|sample url 1> - this is used to describe the bookmark

​_These bookmarks were last updated on 13/03/2022, 14:01:53 using <https://github.com/thearegee/bookworms|Bookworms>_​
",
              "type": "mrkdwn",
            },
            "type": "section",
          },
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Utilties should return all bookmarks - all text 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "blocks": Array [
          Object {
            "text": Object {
              "text": "Bookmarks",
              "type": "plain_text",
            },
            "type": "header",
          },
          Object {
            "type": "divider",
          },
          Object {
            "text": Object {
              "text": "Bookmarks are seperated into different domains, to get the specific bookmarks you can enter the following commands

\`/bookmarks folder 1\`
\`/bookmarks folder 2\`
",
              "type": "mrkdwn",
            },
            "type": "section",
          },
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Utilties should return all bookmarks - blank text 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "blocks": Array [
          Object {
            "text": Object {
              "text": "Bookmarks",
              "type": "plain_text",
            },
            "type": "header",
          },
          Object {
            "type": "divider",
          },
          Object {
            "text": Object {
              "text": "Bookmarks are seperated into different domains, to get the specific bookmarks you can enter the following commands

\`/bookmarks folder 1\`
\`/bookmarks folder 2\`
",
              "type": "mrkdwn",
            },
            "type": "section",
          },
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Utilties should return all bookmarks - no text 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "blocks": Array [
          Object {
            "text": Object {
              "text": "Bookmarks",
              "type": "plain_text",
            },
            "type": "header",
          },
          Object {
            "type": "divider",
          },
          Object {
            "text": Object {
              "text": "Bookmarks are seperated into different domains, to get the specific bookmarks you can enter the following commands

\`/bookmarks folder 1\`
\`/bookmarks folder 2\`
",
              "type": "mrkdwn",
            },
            "type": "section",
          },
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
